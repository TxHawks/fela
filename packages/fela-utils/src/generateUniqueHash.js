/* @flow  */
import hash from 'string-hash'

const collisionBuffer = {}

export default function generateUniqueHash(reference: string | {}): string {
  let result =
    typeof reference === 'object' ? JSON.stringify(reference) : reference
  let prevResult

  do {
    prevResult = result
    result = hash(prevResult).toString(36)
  } while (collisionBuffer[result] && collisionBuffer[result] !== prevResult)

  collisionBuffer[result] = prevResult

  return `_${result}`
}
