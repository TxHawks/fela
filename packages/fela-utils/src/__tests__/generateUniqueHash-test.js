import generateUniqueHash from '../generateUniqueHash'

describe('Generating a unique hash', () => {
  it('should works with objects', () => {
    const reference = {
      color: 'red',
    }

    expect(generateUniqueHash(reference)).toMatchSnapshot()
  })

  it('should not cause a collision between results', () => {
    const reference1 = {
      ':after': {
        content: '"go"',
      },
    }

    const reference2 = {
      ':after': {
        content: '"Fn"',
      },
    }

    expect(generateUniqueHash(reference1)).not.toBe(
      generateUniqueHash(reference2)
    )
  })
})
