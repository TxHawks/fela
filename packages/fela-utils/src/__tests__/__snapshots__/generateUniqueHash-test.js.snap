// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generating a unique hash should works with objects 1`] = `"_137u7ef"`;
